### ISSUE TITLE
Control Globals and functions renaming

### ISSUE THREAD

**Issue by Shine753** [javascript-obfuscator/javascript-obfuscator#629]
I'm a little confused with the readme file.
If I use the renameGlobals=true parameter, it renames globals, and functions. ok.
If I have a function that is called from another script, the call is broken. ok.
So I have to exclude it from the renaming list with reservedNames. ok.

The problem is that doing this, it is not obfuscated.
Is there a way so I can rename it, at obfuscation, without touching the source code ?

For example, saying that this string THISFUNCTION, should always be renamed with this TiHiIiSiFiUiNiCiTiIiOiN or with ghfdffgfhdfjghdfgjdsfhgf when obfuscated.
This way, when I obfuscate any script, I can keep the function call by maintaining some coherence.
Same thing for globals.

note that THISFUNCTION can be a function name, and of course a string corresponding to a javascript call.
```
    SCRIPT 1 : function THISFUNCTION() {}
    SCRIPT 2 : <input type='button' onclick='THISFUNCTION();' />
```

**Comment by sanex3339** [comment]
The solution is to use `name cache` like `uglifyJs` and `terser` are doing.

**Comment by uDelusional** [comment]
Was this ever completed ?

**Comment by sanex3339** [comment]
In the future. Right now I have other priorities.

**Comment by uDelusional** [comment]
Okay, is there a known workaround ?

**Comment by sanex3339** [comment]
only with `reservedNames`

**Comment by uDelusional** [comment]
Returning a Map with the old function/vars => new obf functions/vars would allow us to come up with potential alternatives

**Comment by sanex3339** [comment]
I will implement names cache (that writes into the file) in the future (like uglify.js does) but i haven't any date. Sometime in next year.

**Comment by sanex3339** [comment]
I think, i have to implement this as the next big feature.

**Comment by uDelusional** [comment]
Would be great :)

**Comment by uDelusional** [comment]
I believe this added in yesterday's push ?

**Comment by sanex3339** [comment]
No. I don't know when i will implement this, this spring, i think

**Comment by sanex3339** [comment]
Seems I made a prototype.  Still no tests and documentation.
https://github.com/javascript-obfuscator/javascript-obfuscator/pull/927

The option called `identifierNamesCache` or with CLI `--identifier-names-cache-path`

### With node API:
The option called `identifierNamesCache`.

The option accepts a key=>value object (or an empty object or `null` for initial cache). The default value is `null`.
After obfuscation object with cache can be accessed by `.getIdentifierNamesCache()` method:
```
const obfuscationResult = JavaScriptObfuscator.obfuscate(
        `
            function foo(a, b, c, d) {
              console.log(a, b, c, d)
            }
            
            function bar(...args) {
              foo(...args, 5)
            }
            
            bar(...[1, 2, 3], 4) 
        `,
        {
            ...NO_ADDITIONAL_NODES_PRESET,
            compact: false,
            identifierNamesGenerator: 'mangled',
            renameGlobals: true
        }
);

console.log(obfuscationResult.getIdentifierNamesCache());
```

The output is the following:
```
{ foo: 'a', bar: 'b' }
```



If object with an existing cache is passed, values from the cache will be used for global identifiers:
```
let existingIdentifierNamesCache: TIdentifierNamesCache = {
        bar: 'bar_name_from_cache'
    };
    let obfuscationResult = JavaScriptObfuscator.obfuscate(
        `
            function foo(a, b, c, d) {
              console.log(a, b, c, d)
            }
            
            function bar(...args) {
              foo(...args, 5)
            }
            
            bar(...[1, 2, 3], 4) 
        `,
        {
            ...NO_ADDITIONAL_NODES_PRESET,
            compact: false,
            identifierNamesCache: existingIdentifierNamesCache,
            identifierNamesGenerator: 'mangled',
            renameGlobals: true
        }
    );

console.log(obfuscationResult.getIdentifierNamesCache());
```

The output is the following:
```
{ bar: 'bar_name_from_cache', foo: 'a' }
```


### With CLI:
The option called `--identifier-names-cache-path`
It accepts the path to the existing `.json` file that will be used as cache storage.

Under the hood, it will convert file content to the key=>value dictionary and `identifierNamesCache` option value.
After obfuscation under the hood, it will write the result of `.getIdentifierNamesCache()` back to this file, so the content of this file could be used during the next obfuscation.

The example of file content:
```
{"baz":"a0_0x28d8d0","foo":"a0_0xf337a9","bar":"a0_0x34ba98"}
```

**Comment by sanex3339** [comment]
For more performance if `null` (default value) will be passed to the `identifierNamesCache` - cache will be completely disabled and the result of `.getIdentifierNamesCache()` will be `null`.

If an empty object `{}` or an object with an existing cache will be passed - the result of `.getIdentifierNamesCache()` will be an object with the cache.

**Comment by sanex3339** [comment]
Updated PR with fixes and README documentation.
Still have to refactor some things and write tests.

**Comment by sanex3339** [comment]
Anybody can try this?

**Comment by BlackYuzia** [comment]
hi @sanex3339 
what exact we should test? Just run obfuscation with using `identifierNamesCache` param?
oh, and I should write a small script for manual obfuscate many files?

**Comment by sanex3339** [comment]
You could test how the feature works. Just like in README example or in this example:
```
    SCRIPT 1 : function THISFUNCTION() {}
    SCRIPT 2 : THISFUNCTION();
```

**Comment by BlackYuzia** [comment]
That **should not** work with `module.exports` or `export` etc?
Only if we just obfuscate a few files and run as one script, right?

**Comment by sanex3339** [comment]
With an empty cache object, all global declarations will be added to the cache.
Then this cache may be used to rename all undeclared global variable names in another script.

**Comment by BlackYuzia** [comment]
But if scope is closed ( i mean each file have function that blocks global scope), we can't get access to that function.

I test and get that:
![image](https://user-images.githubusercontent.com/25010528/120103750-dd101180-c159-11eb-85ca-b1dc25afe681.png)

So, that should work only if they are both in the same scope or functions in global scope

**Comment by BlackYuzia** [comment]
```js
// obfuscate.js
const JavaScriptObfuscator = require('javascript-obfuscator');
const config = require('./obf.json');
const fs = require('fs');

const mainSource = fs.readFileSync("./main.js").toString();
const module01Source = fs.readFileSync("./module_01.js").toString();
const module02Source = fs.readFileSync("./module_02.js").toString();

const mainObf = JavaScriptObfuscator.obfuscate(mainSource, config)
const mainCache = mainObf.getIdentifierNamesCache();

config.identifierNamesCache = mainCache; // update cache 

const module01Obf = JavaScriptObfuscator.obfuscate(module01Source, config)
const module01Cache = module01Obf.getIdentifierNamesCache();

config.identifierNamesCache = module01Cache;

const module02Obf = JavaScriptObfuscator.obfuscate(module02Source, config)
const module02Cache = module02Obf.getIdentifierNamesCache();

config.identifierNamesCache = module02Cache;

console.log({
    cache: config.identifierNamesCache,
});

fs.writeFileSync("./dist/main.js", mainObf.getObfuscatedCode());
fs.writeFileSync("./dist/module_01.js", module01Obf.getObfuscatedCode());
fs.writeFileSync("./dist/module_02.js", module02Obf.getObfuscatedCode());
```
```js
// main.js
const gay01 = require('./module_01.js');
const gay02 = require('./module_02.js');

console.log(
    {
        1: gay01(),
        2: gay02(),
    }
);
```
```js
// module_01.js
function module01Gay() { return true };

module.exports = module01Gay;
```
```js
// module_02.js
function module02Gay() { return true };

module.exports = module02Gay;
```
```js
// obf.json (config)
{
    "compact": false,
    "controlFlowFlattening": true,
    "controlFlowFlatteningThreshold": 1.0,
    "deadCodeInjection": true,
    "deadCodeInjectionThreshold": 1,
    "debugProtection": false,
    "debugProtectionInterval": false,
    "disableConsoleOutput": false,
    "domainLock": [],
    "forceTransformStrings": [],
    "identifierNamesGenerator": "hexadecimal",
    "identifiersDictionary": [],
    "identifiersPrefix": "",
    "ignoreRequireImports": false,
    "inputFileName": "",
    "log": true,
    "numbersToExpressions": true,
    "optionsPreset": "high-obfuscation",
    "renameGlobals": true,
    "renameProperties": true,
    "renamePropertiesAutoExclude": true,
    "reservedNames": [],
    "reservedStrings": [],
    "rotateStringArray": true,
    "seed": 0,
    "selfDefending": true,
    "shuffleStringArray": true,
    "simplify": true,
    "sourceMap": false,
    "sourceMapBaseUrl": "",
    "sourceMapFileName": "",
    "sourceMapMode": "separate",
    "splitStrings": true,
    "splitStringsChunkLength": 10,
    "stringArray": true,
    "stringArrayCallsIndexType": [
        "hexadecimal-number"
    ],
    "stringArrayEncoding": [
        "rc4"
    ],
    "stringArrayIndexShift": true,
    "stringArrayWrappersCount": 10,
    "stringArrayWrappersChainedCalls": true,
    "stringArrayWrappersParametersMaxCount": 10,
    "stringArrayWrappersType": "function",
    "stringArrayThreshold": 1,
    "target": "browser",
    "transformObjectKeys": true,
    "unicodeEscapeSequence": false,
    "identifierNamesCache": {}
}
```

### Results 

**Without obfuscate**
![image](https://user-images.githubusercontent.com/25010528/120103885-78a18200-c15a-11eb-86e2-3336bd9c8469.png)

**With try to obfuscate**
![image](https://user-images.githubusercontent.com/25010528/120103937-bb635a00-c15a-11eb-9d20-9b14c73f0eed.png)

**Cache**
```js
  cache: {
    _0x4f6f5c: '_0x25b7c7',
    _0x37d61f: '_0x7bd350',
    gay01: '_0x19ceee',
    gay02: '_0x3e9e46',
    _0x1bda99: '_0x8b4f96',
    _0x32b478: '_0xa89b5b',
    _0x5d688f: '_0xaab0de',
    module01Gay: '_0x5a9418',
    _0x1d786a: '_0x123023',
    _0x35da71: '_0x512db4',
    module02Gay: '_0x4ac5a4'
  }
  ```

**Comment by BlackYuzia** [comment]
oh wait. Did that function just save variables for us to replace manualy?

**Comment by sanex3339** [comment]
No. These cached names can be passed again as `identifierNamesCache` value. All this stuff will save to the cache ONLY global identifier names. 
And ONLY global identifier names WITHOUT declaration will be renamed with the values from the cache if the existing cache is passed.

**Comment by sanex3339** [comment]
All stuff that you described above won't work, because this is not the case for this feature. This feature mainly for browser code that may be separated between multiple files.

```
// main.js
const module1 = require('./module_01.js');
const module2 = require('./module_02.js');

console.log(
    {
        1: module1(),
        2: module2(),
    }
);
```

This won't work because both module variable names are declared.

**Comment by BlackYuzia** [comment]
@sanex3339 what about exported and imported modules?
I mean support obfuscate multiple files and etc.
What about that? 🤔

**Comment by sanex3339** [comment]
What problem with multiple module files?

**Comment by BlackYuzia** [comment]
We can't obfuscate a few files in the same time.
I mean, if we obfuscate module and main file they don't work as well.

**Comment by sanex3339** [comment]
Can you show a small example?

**Comment by BlackYuzia** [comment]
```js
// obfuscate.js
const JavaScriptObfuscator = require('javascript-obfuscator');
const config = require('./obf.json');
const fs = require('fs');

const mainSource = fs.readFileSync("./main.js").toString();
const module01Source = fs.readFileSync("./module_01.js").toString();
const module02Source = fs.readFileSync("./module_02.js").toString();

const module01Obf = JavaScriptObfuscator.obfuscate(module01Source, config)
const module02Obf = JavaScriptObfuscator.obfuscate(module02Source, config)
const mainObf = JavaScriptObfuscator.obfuscate(mainSource, config)

fs.writeFileSync("./dist/main.js", mainObf.getObfuscatedCode());
fs.writeFileSync("./dist/module_01.js", module01Obf.getObfuscatedCode());
fs.writeFileSync("./dist/module_02.js", module02Obf.getObfuscatedCode());
```

```js
// main.js
const localObject = { a: 1, b: 2, c: 3 };         
const importedArray = require("./module_01.js");  

console.log(
    localObject,
    importedArray
);
```

```js
// module_01.js
module.exports = [
    0, 1, 2, 34, 5
]
```
```js
// config the same as above
```

```js
// without obfuscate
// $ node main.js 
{ a: 1, b: 2, c: 3 } [ 0, 1, 2, 34, 5 ]
```
```js
// with obfuscation
// $ node dist/main.js
{ a: 1, b: 2, c: 3 } {}
```

**Comment by sanex3339** [comment]
This is because `renameProperties` option is enabled. Don't use it.
It converts
`module.exports =` to the `module['_0x14b6fd'] = `

**Comment by BlackYuzia** [comment]
yep, I mean, if we have cache. We can use it in this case, no?

**Comment by sanex3339** [comment]
yep, I mean, if we have a cache. We can use it in this case, no?
Hmm. Maybe, I'll think about property names. Anyway, in the case with `exports` this won't help, because `import` in `main` module looks like: `importedArray=require(_0x49d4c9(0x131));`, so nothing about `exports` field.

**Comment by BlackYuzia** [comment]
also, we have second trouble. I suppose we can fix it by cache names. Mb we talk about that trouble, I don't remember actualy

```js
    // config
    "renameGlobals": true,
    "renameProperties": true,
```
```js
// main.js
const Kruzya = require("./module_01.js");
const kruzya = new Kruzya("Kruzya");
console.log({ kruzya, randomGetter: kruzya.randomGetter, fieldRandomName: kruzya.fieldRandomName });
```
```js
// module.js
class Kruzya {
    constructor(fieldRandomName) {
        this.fieldRandomName = fieldRandomName || null;
    }

    get randomGetter() {
        return true;
    }

    set randomSetter(v) {
        this.fieldRandomName = v;
        return this;
    }
}

module.exports = Kruzya;
```

Result 
![image](https://user-images.githubusercontent.com/25010528/120106783-1dc25780-c167-11eb-902f-1dd4abccb3ac.png)

```js
// obfuscated main.js
const _0x53e7d2 = require('./module_01.js');
const _0x285545 = new _0x53e7d2('Kruzya');
console['log']({
    '_0x574842': _0x285545,
    '_0x3d9cf5': _0x285545['_0x3d9cf5'],
    '_0x214c4b': _0x285545['_0x214c4b']
});
```
```js
// obfuscated module.js
class _0x31ac8a {
    constructor(_0x23265f) {
        this['_0x1ad7a1'] = _0x23265f || null;
    }
    get ['_0x42e22b']() {
        return !![];
    }
    set ['_0x2964f3'](_0x4e2562) {
        this['_0x1ad7a1'] = _0x4e2562;
        return this;
    }
}
module['exports'] = _0x31ac8a;
```

fix (in my mind): add to cache `_0x1ad7a1` variable, then export cache to main.js before obfuscate and use `_0x1ad7a1` as replace  and etc... The same with `_0x42e22b` and `_0x2964f3` variables. 

That things are simple, I think :eyes:

**Comment by sanex3339** [comment]
Yes, this is the same. Property names cache can fix this. I will think about how I can add property name cache.

**Comment by sanex3339** [comment]
Probably i will rework current cache structure from the plain key=>value dictionary to the following:

{
    globalIdentifiers: {} // key=>value,
    propertyIdentifiers: {} // key=>value
}

**Comment by sanex3339** [comment]
Updated PR with properties cache. @BlackYuzia please try it with your previous examples.

**Comment by BlackYuzia** [comment]
@sanex3339 give me some advice. How I could use `identifierNamesCache`, `globalIdentifiers` and `propertyIdentifiers` now?
I try use `identifierNamesCache`, but get error:
```js
ReferenceError: Validation failed. errors:
`identifierNamesCache` errors:
    - Passed value must be an identifier names cache object or `null` value
```
```js
    // config
    "identifierNamesCache": {},
    "globalIdentifiers": {},
    "propertyIdentifiers": {}
```

**Comment by sanex3339** [comment]
Oh, seems this is a bug

**Comment by BlackYuzia** [comment]
hmm, that means what we can do something like REPLACE function inside obfuscator?

Like:
```js
// config
{
"identifierNamesCache": { _date: new Date() } // just example, mb I wrong 
}
// code 
const date = _date; // _date is not defined, but should be replaced by "new Date()"
```

**Comment by sanex3339** [comment]
With 
```
"identifierNamesCache": {
     "globalIdentifiers": {
         _date: new Date() 
    }
}
```

You can do this, but this is not the expected use case for this option (because I limited `globalIdentifiers` usage to undeclared variables.
If sometime in the future I will add the way to control identifier names, it will the way through `identifierNamesGenerator` option.

**Comment by BlackYuzia** [comment]
Tested, that work now
![image](https://user-images.githubusercontent.com/25010528/120177321-fde76e00-c210-11eb-9d3d-a8cfffdf3a10.png)

**Comment by sanex3339** [comment]
The empty option object now should work correctly

**Comment by BlackYuzia** [comment]
```js
ReferenceError: Validation failed. errors:
`identifierNamesCache` errors:
    - Passed value must be an identifier names cache object or `null` value
```
```js
    "identifierNamesCache": {
        "globalIdentifiers": {
            "KRUZYAPIDOR": true
        },
        "propertyIdentifiers": {}
    }
```
did that bug? :eyes:

**Comment by sanex3339** [comment]
No. Dictionary values should be strings.

**Comment by BlackYuzia** [comment]
![image](https://user-images.githubusercontent.com/25010528/120182952-bca68c80-c217-11eb-9c2b-33fb9d22ca04.png)
yep, work now. Thx

**Comment by BlackYuzia** [comment]
@sanex3339 for use this like `replace in code`, we should add second obfuscate (or obfuscate only replaced values) or we get unobfuscated values.

**Comment by BlackYuzia** [comment]
yep, I now. I mean, if we add feature for replace code as **expected**, we also should add obfuscation for replaced values.

**Comment by BlackYuzia** [comment]
or we have second choice. 
obfuscated values before replace.

**Comment by sanex3339** [comment]
I don't know how i will do this right now. Right now this is the cache only for reuse some names between files

**Comment by sanex3339** [comment]
Thank you for testing. Probably from tomorrow and all next week I won't be able to finish this PR, because of my vacation. I'll finish in a week.

**Comment by sanex3339** [comment]
Why not? Released as `2.14.0`. Enjoy!
